src/state.rs

```diff
@@ -109,6 +111,244 @@ pub struct ResolvedDns {
     dns_refresh_rate: std::time::Duration,
 }
 
+/// A SingleFlight can make sure that only the first request to really handle DNS resolving task
+/// for specified hostname and other requests in concurrency need to wait for the task completion.
+#[derive(serde::Serialize, Default, Debug)]
+struct SingleFlight {
+    // is_first_req_in can record requests number and it can be used to
+    // check the request number, such as the first request.
+    is_first_req_in: bool,
```

Comment:
I think I mentioned in a previous review, but I don't think you need this flag since we have a lock around this structure. Instead just lock, check if a single flight already exists, if so use it, if not add a new one. Does that work?

Reply:
Hi @nmittler, this flag should be here because `checking the index of the request` and `doing the dns resolving task` are not at the same time. When multiple requests get into `resolve_host` in concurrency, the flag must be set to distinguish the first request or other request.

Reply:
done after the discussion.